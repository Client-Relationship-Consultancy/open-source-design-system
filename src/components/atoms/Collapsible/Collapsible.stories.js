import React from "react"

import { storiesOf } from "@storybook/react"
import { boolean } from "@storybook/addon-knobs"

import { Collapsible } from "../../../index"

const stories = storiesOf("Collapsible", module)

const collapsedContent = (
  <>
    <h1>Panel</h1>
    <p>This is some text to help explain what this panel is.</p>
    <p>
      <em>Click the arrow to show the hidden content...</em>
    </p>
  </>
)

const openContent = (
  <>
    <p>This is the hidden content you seek!</p>
    <p>This is the hidden content you seek!</p>
    <p>This is the hidden content you seek!</p>
    <p>This is the hidden content you seek!</p>
    <p>This is the hidden content you seek!</p>
  </>
)

stories.add("Collapsible", () => {
  const open = boolean("Open", false)
  const locked = boolean("Locked", false)

  return (
    <Collapsible
      open={open}
      locked={locked}
      collapsedContent={collapsedContent}
      clickHandler={() => console.log("clicked!")}
    >
      {openContent}
    </Collapsible>
  )
})
