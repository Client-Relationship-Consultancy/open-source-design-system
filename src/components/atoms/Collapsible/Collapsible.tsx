import React from "react";
import styled from "styled-components";

import Icon from "../Icon";

const StyledButton = styled.button`
  position: absolute;
  top: 0;
  right: 0;
  background-color: transparent;
  border: 0;
  outline: none;
  cursor: pointer;
`;

const CollapsiblePanel = styled.div`
  position: relative;
  font-size: 1rem;
  min-height: 2em;
  width: 100%;
`;

CollapsiblePanel.displayName = "CollapsiblePanel";

interface ICollapsibleIconProps {
  open: boolean;
}

const CollapsibleIcon = styled(Icon)<ICollapsibleIconProps>`
  svg {
    transform: ${({ open }) => (open ? "rotateZ(180deg)" : "rotateZ(0deg)")};
    transition: transform 0.5s;
  }
`;

const CollapsedContentContainer = styled.div``;

CollapsedContentContainer.displayName = "CollapsedContentContainer";

interface IOpenContentContainerProps {
  open: boolean;
}

const OpenContentContainer = styled.div<IOpenContentContainerProps>`
  visibility: ${({ open }) => (open ? "visible" : "hidden")};
  height: ${({ open }) => (open ? "100%" : "0%")};
  opacity: ${({ open }) => (open ? 1 : 0)};
  transition-duration: 0s, 1s, 1s;
  transition-delay: ${({ open }) => (open ? "0s" : "1s")}, 0s, 0s;
  transition-property: visibility, height, opacity;
`;

OpenContentContainer.displayName = "OpenContentContainer";

interface ICollapsibleProps {
  open: boolean;
  locked?: boolean;
  children: React.ReactNode;
  collapsedContent?: React.ReactNode;
  width?: number;
  index: number;
  clickHandler: (event: React.MouseEvent<HTMLButtonElement>) => void;
}

const Collapsible: React.FC<ICollapsibleProps> = (props) => {
  const { locked = false, open, children, collapsedContent, clickHandler, index } = props;

  return (
    <CollapsiblePanel>
      <StyledButton type="button" data-index={index} onClick={clickHandler}>
        <CollapsibleIcon name="chevron-down" size="1x" open={open} />
      </StyledButton>
      {collapsedContent && (
        <CollapsedContentContainer>{collapsedContent}</CollapsedContentContainer>
      )}
      <OpenContentContainer open={open}>{children}</OpenContentContainer>
    </CollapsiblePanel>
  );
};

Collapsible.displayName = "Collapsible";

export default Collapsible;
