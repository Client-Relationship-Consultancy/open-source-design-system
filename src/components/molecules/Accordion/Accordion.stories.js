import React from "react"

import { storiesOf } from "@storybook/react"

import { Accordion } from "../../../index"

const stories = storiesOf("Accordion", module)

const openContent = (
  <div style={{ padding: "0.5rem 1rem" }}>
    <h2 style={{ marginTop: "0px" }}>Open Content</h2>
    <p>This is some high quality open content</p>
    <p>It&apos;s even got a button:</p>
    <button type="button" onClick={() => null}>
      I don&apos;t actually do anything though :)
    </button>
  </div>
)

const collapsedContent = (
  <div style={{ padding: "0.5rem 1rem" }}>
    <h1>Collapsed Content</h1>
    <p>If you click the arrow on the right, you&apos;ll see the hidden content!</p>
  </div>
)

const content = [
  {
    collapsedContent,
    openContent,
  },
  {
    collapsedContent,
    openContent,
  },
  {
    collapsedContent: (
      <div style={{ padding: "0.5rem 1rem" }}>
        <h1>Collapsed Content</h1>
        <p>Clicking the arrow won&apos;t open me unless you unlock me first</p>
      </div>
    ),
    openContent,
  },
  {
    collapsedContent: undefined,
    openContent,
  },
]

class MockAccordionParent extends React.Component {
  state = {
    openPanels: [0],
    lockedPanels: [2],
  }

  updateOpenAccordionPanels = (updatedOpenPanels) => {
    this.setState({ openPanels: updatedOpenPanels })
  }

  render() {
    return (
      <Accordion
        open={this.state.openPanels}
        locked={this.state.lockedPanels}
        content={content}
        updateOpenPanels={this.updateOpenAccordionPanels}
      />
    )
  }
}

stories.add("Accordion with three panels", () => {
  return <MockAccordionParent />
})
