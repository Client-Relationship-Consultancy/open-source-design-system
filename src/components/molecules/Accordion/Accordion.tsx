import React from "react";
import styled from "styled-components";
import { colourPalette } from "../../../brandColours";

import Collapsible from "../../atoms/Collapsible";

type AccordionContent = { collapsedContent: React.ReactNode; openContent: React.ReactNode };

interface IAccordionContainerProps {
  width?: number;
}

const AccordionContainer = styled.div<IAccordionContainerProps>`
  width: ${({ width }) => (width ? `${width}rem` : "100%")};
  background-color: ${({ theme }) => theme.background.hex};
`;

AccordionContainer.defaultProps = {
  theme: colourPalette.examplePalette,
};

interface IAccordionProps {
  open: number[];
  locked: number[];
  updateOpenPanels: (updatedOpenPanels: number[]) => void;
  content: AccordionContent[];
  width?: number;
}

class Accordion extends React.Component<IAccordionProps> {
  handleOnClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    const { open, locked } = this.props;

    const updatedOpen = [...open];

    if (event.currentTarget.dataset.index) {
      const index = parseInt(event.currentTarget.dataset.index, 10);

      if (!locked.includes(index)) {
        const existingIndex = updatedOpen.findIndex((val) => val === index);
        if (existingIndex > -1) {
          updatedOpen.splice(existingIndex, 1);
        } else {
          updatedOpen.push(index);
        }
      }
    }

    return updatedOpen;
  };

  renderPanels = (): JSX.Element[] => {
    const { content, updateOpenPanels, open = [], locked = [] } = this.props;

    const panels = content.map((item: AccordionContent, index: number) => (
      <Collapsible
        /* eslint-disable react/no-array-index-key */
        key={`collapsible-${index}`}
        index={index}
        open={open.includes(index)}
        locked={locked.includes(index)}
        collapsedContent={item.collapsedContent}
        clickHandler={(event) => {
          updateOpenPanels(this.handleOnClick(event));
        }}
      >
        {item.openContent}
      </Collapsible>
    ));

    return panels;
  };

  render(): JSX.Element {
    const { width } = this.props;

    return <AccordionContainer width={width}>{this.renderPanels()}</AccordionContainer>;
  }
}

export default Accordion;
