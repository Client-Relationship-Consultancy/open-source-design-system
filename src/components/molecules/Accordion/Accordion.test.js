import React from "react"
import { shallow } from "enzyme"
import Accordion , { accordionWidthCalculator } from "./Accordion"
import Collapsible from "../../atoms/Collapsible"

const mockContent = [
  {
    collapsedContent: <h1>Mock Collapsed Content 1</h1>,
    openContent: <p>Mock open content 1</p>,
  },
  {
    collapsedContent: <h1>Mock Collapsed Content 2</h1>,
    openContent: <p>Mock open content 2</p>,
  },
  {
    collapsedContent: <h1>Mock Collapsed Content 3</h1>,
    openContent: <p>Mock open content 3</p>,
  },
]

const mockUpdateOpenPanels = jest.fn()

describe("Accordion", () => {
  it("should match last snaphot when default open & closed used", () => {
    const component = shallow(
      <Accordion content={mockContent} updateOpenPanels={mockUpdateOpenPanels} />,
    )
    expect(component).toMatchSnapshot()
  })

  it("test helper functions for style",()=>{
    expect(accordionWidthCalculator({ width: 10 })).toBe("10rem")
    expect(accordionWidthCalculator({ })).toBe("100%")
  })

  it("should return the correct open panel array when the handleOnClick is triggered", () => {
    const component = shallow(
      <Accordion
        open={[0]}
        locked={[2]}
        content={mockContent}
        updateOpenPanels={mockUpdateOpenPanels}
      />,
    )
    let result = component.instance().handleOnClick({ currentTarget: { dataset: { index: "0" } } })
    expect(result).toEqual([])

    result = component.instance().handleOnClick({ currentTarget: { dataset: { index: "1" } } })
    expect(result).toEqual([0, 1])

    result = component.instance().handleOnClick({ currentTarget: { dataset: { index: "2" } } })
    expect(result).toEqual([0])

    result = component.instance().handleOnClick({ currentTarget: { dataset: {  } } })
    expect(result).toEqual([0])
  })
  it("should render the the collapsibles correctly based on the content passed", () => {
    const component = shallow(
      <Accordion
        open={[0]}
        locked={[2]}
        content={mockContent}
        updateOpenPanels={mockUpdateOpenPanels}
      />,
    )
    const collapsibles = component.instance().renderPanels()
    expect(collapsibles).toHaveLength(3)

    expect(component.find(Collapsible).get(0).props.open).toBe(true)
    expect(component.find(Collapsible).get(0).props.locked).toBe(false)
    expect(component.find(Collapsible).get(0).props.index).toBe(0)
    expect(component.find(Collapsible).get(0).props.collapsedContent).toBe(
      mockContent[0].collapsedContent,
    )
    expect(component.find(Collapsible).get(0).props.children).toBe(mockContent[0].openContent)

    expect(component.find(Collapsible).get(1).props.open).toBe(false)
    expect(component.find(Collapsible).get(1).props.locked).toBe(false)
    expect(component.find(Collapsible).get(1).props.index).toBe(1)
    expect(component.find(Collapsible).get(1).props.collapsedContent).toBe(
      mockContent[1].collapsedContent,
    )
    expect(component.find(Collapsible).get(1).props.children).toBe(mockContent[1].openContent)

    expect(component.find(Collapsible).get(2).props.open).toBe(false)
    expect(component.find(Collapsible).get(2).props.locked).toBe(true)
    expect(component.find(Collapsible).get(2).props.index).toBe(2)
    expect(component.find(Collapsible).get(2).props.collapsedContent).toBe(
      mockContent[2].collapsedContent,
    )
    expect(component.find(Collapsible).get(2).props.children).toBe(mockContent[2].openContent)
  })
})
