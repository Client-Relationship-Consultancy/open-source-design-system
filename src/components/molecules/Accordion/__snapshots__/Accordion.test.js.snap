// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should match last snaphot when default open & closed used 1`] = `
<Accordion__AccordionContainer>
  <Collapsible
    clickHandler={[Function]}
    collapsedContent={
      <h1>
        Mock Collapsed Content 1
      </h1>
    }
    index={0}
    key="collapsible-0"
    locked={false}
    open={false}
  >
    <p>
      Mock open content 1
    </p>
  </Collapsible>
  <Collapsible
    clickHandler={[Function]}
    collapsedContent={
      <h1>
        Mock Collapsed Content 2
      </h1>
    }
    index={1}
    key="collapsible-1"
    locked={false}
    open={false}
  >
    <p>
      Mock open content 2
    </p>
  </Collapsible>
  <Collapsible
    clickHandler={[Function]}
    collapsedContent={
      <h1>
        Mock Collapsed Content 3
      </h1>
    }
    index={2}
    key="collapsible-2"
    locked={false}
    open={false}
  >
    <p>
      Mock open content 3
    </p>
  </Collapsible>
</Accordion__AccordionContainer>
`;
